### ISSUE TITLE
Built-in collectors fail promtool linting

### ISSUE THREAD

**Issue by mpalmer** [deadtrickster/prometheus.erl#135]
With a stock install of `prometheus.ex` exporting the default Erlang metrics, the following linting failures are reported:

```
$ wget -O - -q http://localhost:9548/metrics | promtool check metrics
erlang_vm_statistics_context_switches counter metrics should have "_total" suffix
erlang_vm_statistics_garbage_collection_bytes_reclaimed counter metrics should have "_total" suffix
erlang_vm_memory_atom_bytes_total non-counter metrics should not have "_total" suffix
erlang_vm_statistics_garbage_collection_number_of_gcs counter metrics should have "_total" suffix
erlang_vm_statistics_garbage_collection_words_reclaimed counter metrics should have "_total" suffix
erlang_vm_statistics_runtime_milliseconds counter metrics should have "_total" suffix
erlang_vm_statistics_wallclock_time_milliseconds counter metrics should have "_total" suffix
erlang_vm_memory_bytes_total non-counter metrics should not have "_total" suffix
erlang_vm_memory_processes_bytes_total non-counter metrics should not have "_total" suffix
erlang_vm_memory_system_bytes_total non-counter metrics should not have "_total" suffix
erlang_vm_statistics_run_queues_length_total non-counter metrics should not have "_total" suffix
erlang_vm_atom_count non-histogram and non-summary metrics should not have "_count" suffix
erlang_vm_port_count non-histogram and non-summary metrics should not have "_count" suffix
erlang_vm_process_count non-histogram and non-summary metrics should not have "_count" suffix
erlang_vm_statistics_runtime_milliseconds use base unit "seconds" instead of "milliseconds"
erlang_vm_statistics_wallclock_time_milliseconds use base unit "seconds" instead of "milliseconds"
```

Allowing people to adjust their base units if they like is nice, but IMAO the out-of-the-box defaults should be `promtool` compliant.

**Comment by NelsonVides** [comment]
I really wonder, how critical this would be. I want to release a new major version, with breaking changes, so if we're to change metric names this would be the perfect chance, but on the other hand, it'd break a lot of existing dashboards with metrics that don't exist anymore. Looking for opinions @mkuratczyk @onno-vos-dev @mikpe @deadtrickster (tagging a few random members).

**Comment by onno-vos-dev** [comment]
I'm a bit torn on this sort of change for the simple purpose of making a linter happy. The current metric names are descriptive and simply changing them and breaking all existing dashboards for this purpose seems a bit excessive in my opinion would be my initial $0,02. 

I'd agree that out-of-the-box defaults should be promtool compliant but how important is this really? Is a happy linter more important than existing dashboards?
Does `promtool check metrics` not have a way to disable the linter? It doesn't look like it based on [the documentation](https://prometheus.io/docs/prometheus/latest/command-line/promtool/#promtool-check-metrics) at least? 

I'm a bit torn on this...

**Comment by mikpe** [comment]
I agree that existing dashboards are more important than satisfying that linter. OTOH I don't want to stand in the way of DTRT.

If there was a toogle to select new or old behaviour then I think a breaking update with that toogle prominently documented could be Ok.

**Comment by paulo-ferraz-oliveira** [comment]
A major is a major, right? As long as there's a guide that explains clearly, what it'll break nobody's forcing consumers. I also want to DTRT, and am also torn on this one; maybe one good thing we could do would be to integrate `promtool` in this repo's CIs to at least prevent further non-compliance (?)

**Comment by NelsonVides** [comment]
I've tried to add `promtool` to CI and for the love of everything dear GitHub Actions are making me feel stupid (https://github.com/prometheus-erl/prometheus.erl/pull/188) 😅 promtool was actually easy to install but holy crap I can't get Erlang to print the format into anything I can pipe from 🤦🏽‍♂️

I'm also not convinced about breaking existing dashboards, that would be a huge PITA. Also, services like DataDog charge for the number of metrics, so changing the metric name would mean a user would be charged for two metrics now, the old name and the new one, if he wants to keep the history. Somebody could get really angry at this if you then include cardinality from a large number of nodes for example. Nobody's forced to upgrade to a major but, we're not backporting improvements to older versions so that'd mean users would be forced to choose between breaking dashboards and improvements everywhere else.

But also... promtool changed requirements when Prometheus was integrated into OpenMetrics. And OpenMetrics 2.0 seems to be planning to change many of these conventions (https://github.com/orgs/prometheus/projects/11/views/1), so well, maybe one day we'll just want to implement OM2.0 compatibility at once 🤷🏽‍♂️

I'd only wish we could at least have promtool for any future new metrics created. But promtool also doesn't have an easy way to ignore bad names 😐

Dunno, I'd close this as "won't do" and only hope we reviewers remember to manually check any new metric that gets proposed through. What do you guys think?

**Comment by paulo-ferraz-oliveira** [comment]
You do have good points there:

1. breaking stuff is a PITA
2. paying for cardinality (e.g., in Datadog) is a bummer
3. maybe the future will say something different about these names

I wouldn't mind not moving this forward, but you can also leave it open for "when that day comes". (sometimes it's just about waiting)

Otoh, I can't really understand what the issues with Erlang are, but if you give me an example of what's not working maybe if I have sometime I can give it a jab (fresh look and all that)...

**Comment by NelsonVides** [comment]
@paulo-ferraz-oliveira 

The issue is that we'd need to put the output of `prometheus_text_format:format/1` into a file that then we can pipe into `promtool`

The following works on my computer:
```erlang
    rebar3 shell --eval 'ok = file:write_file("prom_format.txt", prometheus_text_format:format()), halt().'
    cat prom_format.txt | promtool check metrics
```

But on github actions it does not even print anything to IO (including any helpers like cwd or the contents of the format or anything else), and `cat` fails forever with the file not existing.

**Comment by mikpe** [comment]
I rarely use `rebar3 shell`, but plain Erlang does let you write normal command-line tools that print to standard output without additional garbage, so perhaps try something like

`erl -pa ... -noshell -eval 'ok = io:format(standard_io, "~s~n", [prometheus_text_format:format()]), halt().' | promtool check metrics`

**Comment by NelsonVides** [comment]
@mikpe yes, thank you, that helped! https://github.com/prometheus-erl/prometheus.erl/actions/runs/16413437748/job/46373727553 😄 

Ok, gonna try to give that PR some sense, maybe there's _some_ checking we can introduce into CI 🤔

**Comment by NelsonVides** [comment]
Ok, now it works. These are the metrics that would change name if we wanted to merge promtool in a green state: https://github.com/prometheus-erl/prometheus.erl/commit/ae32afd2a726bfe23f85526bd0d253f799a5f6f2

See https://github.com/prometheus-erl/prometheus.erl/pull/188 for details.

Now we have four options:
1. We drop as won't do.
2. We merge and mercilessly introduce all the breaking changes in metric names.
3. We introduce a way to ignore from the promtool check the already broken metric names.
4. We introduce a feature flag that would use one metric name or another for at least one major (or at least a quarter, 3 months, or some other release strategy, you tell me).

Thoughts? :)

**Comment by paulo-ferraz-oliveira** [comment]
I vote 3 and 4. I don't like to maintain legacy code (especially in FOSS), but sometimes it's required. I also don't like introducing (needless?) breaking changes but sometimes it's required. How much of a "standard" is that tool?

Another alternative would be to allow metric names to be changed e.g., via a callback, for those wanting to keep compatibility. But at least for new consumers that wouldn't be a problem (?)

**Comment by mikpe** [comment]
I vote 4.

**Comment by onno-vos-dev** [comment]
I vote 4 and we leave the user with the choice 👍 

Complete order of my preference would be 4, 3, 1, 2.

**Comment by NelsonVides** [comment]
Ok, done, point n4, wasn't too complicated after all. Please have a look at https://github.com/prometheus-erl/prometheus.erl/pull/188 🔥
